rfc-079-script-imports-of-production-data.md

```diff
@@ -0,0 +1,67 @@
+# Script imports of production data
+
+## Summary
+
+Developers writing production data to file on a local or remote server MUST do
+this via a script. It MUST NOT be done by manual actions, such as via the Rails
+console or `pg_dump`.
+
+This script MUST anonymise any sensitive data, and the data dump MUST be stored
+outside of any Git repositories to ensure it can't be inadvertently committed
+alongside any changes to the codebase.
+
+## Problem
+
+Fetching data from a database by hand is error-prone. A data dump may be left
+on a server, committed to a repository or even in some cases alter production
+data in unforeseen ways. Developers may not realise they have exposed or
+impacted production data until long after an incident occurs.
+
+By writing a script that developers can run from their local machine to
+automate this process, we ensure that every import of data from a production
+environment is done in the same way, minimising room for error. It also allows
+us to scrub the resulting data dump of any personally-identifying and
+security-sensitive information.
+
+Committing a data-fetching script to the repository makes it visible to other
+members of the team in the code review process, allows it to be developed
+against a local instance of the database, and allows others to provide feedback
+to ensure we're reaching a consensus on the safest way of fetching this data.
+
+## Proposal
+
+If developing an application requires a developer to write any production data
+containing sensitive data to a local file or a file on a remove server, this
+MUST be done via a script.
+
+The script MUST anonymise any security-sensitive data. It SHOULD remove or
+anonymise any personal data where relevant. Anonymisation SHOULD be done
+before any data is written to disc. This MAY be done on the remote server, or
+piped through a script that substitutes any sensitive values before writing to
+a file.
+
+The script MUST have gone through the usual code review process with
+other members of the team before it is run against a production database.
+
+The file written to by the script MUST be stored outside any Git repositories,
+for example in a `/tmp/dumps/projectname.sql` directory.
+
+The file written to by the script MUST be temporary and SHOULD be deleted
+within 30 days of creation.
```

Comment:
I think it's probably wiser/safer for the script to do, the dump, import the data and then throw away the dump once it's imported. This is [what I did in the script I wrote for BEIS](https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/blob/develop/bin/db-restore)

Reply:
That scares me, too, but as long as it's all anonymized, I guess it's ok. Might be worth elevating to MUSTs for anonymization if the data is being written to a non-production database.

On this point, though. I don't think that's the only case for dumping data from production (see Dom's eLinks example above), but it's a sensible suggestion if it applies.

Reply:
I've added James's suggestion to store the file in `tmp` so it will be automatically removed within 3 days (or equivalent) and if not to ensure this is deleted within 30 days. Do you think that's enough to cover this case?

Reply:
I think there's something about using production data in a non-production database that means anonymization needs to be a MUST if that's happening.

Reply:
Added a suggestion to cover this.